infra/test/Test/Pos/Infra/Arbitrary/Ssc.hs

```diff
@@ -0,0 +1,282 @@
+    } deriving Generic
+
+----------------------------------------------------------------------------
+-- Core
+----------------------------------------------------------------------------
+
+-- | Wrapper over 'SignedCommitment'. Creates an invalid SignedCommitment
+-- w.r.t. 'verifyCommitmentSignature'.
+newtype BadSignedCommitment = BadSignedComm
+    { getBadSignedC :: SignedCommitment
+    } deriving (Generic, Show, Eq)
+
+instance Arbitrary BadSignedCommitment where
+    arbitrary = BadSignedComm <$> do
+        pk <- arbitrary
+        sig <- arbitrary
+        comm <- Commitment <$> arbitrary <*> arbitrary
+        return (pk, comm, sig)
+    shrink = genericShrink
+
+-- | Pair of 'Commitment' and 'Opening'.
+data CommitmentOpening = CommitmentOpening
+    { coCommitment :: !Commitment
+    , coOpening    :: !Opening
+    } deriving (Generic, Show)
+
+-- | Wrapper over '(Commitment, Opening)'. Creates an invalid pair of a Commitment and an
+-- Opening w.r.t. 'verifyOpening'.
+data BadCommAndOpening = BadCommAndOpening
+    { getBadCAndO :: (Commitment, Opening)
+    } deriving (Generic, Show, Eq)
+
+instance Arbitrary BadCommAndOpening where
+    arbitrary = do
+        badComm <- Commitment <$> arbitrary <*> arbitrary
+        opening <- arbitrary
+        return $ BadCommAndOpening (badComm, opening)
+    shrink = genericShrink
+
+-- | Generate 50 commitment/opening pairs for tests.
+commitmentsAndOpenings :: [CommitmentOpening]
+commitmentsAndOpenings =
+    map (uncurry CommitmentOpening) $
+    deterministic "commitmentsAndOpenings" $ replicateM 50 $ do
+      t <- randomNumberInRange 3 10
+      n <- randomNumberInRange (t*2-1) (t*2)
+      vssKeys <- replicateM (fromInteger n) $ toVssPublicKey <$> vssKeyGen
+      randCommitmentAndOpening (fromIntegral t) (NE.fromList vssKeys)
+
+instance Arbitrary CommitmentOpening where
+    arbitrary = elements commitmentsAndOpenings
+    shrink = genericShrink
+
+instance Nonrepeating CommitmentOpening where
+    nonrepeating n = sublistN n commitmentsAndOpenings
+
+instance Arbitrary Commitment where
+    arbitrary = coCommitment <$> arbitrary
+    -- No other field is shrunk in the implementation of 'shrink'
+    -- for this type because:
+    -- 1. The datatype's invariant cannot be broken
+    -- 2. The cryptographic datatypes used here don't have 'shrink' implemented
+    shrink Commitment {..} = [ Commitment { commShares = shrunkShares, .. }
+                             | shrunkShares <- filter (not . null) $ shrink commShares
+                             ]
+
+-- | A 'CommitmentsMap' is made from a '[SignedCommitment]', where
+-- 'SignedCommitment' is a type alias for
+-- '(PublicKey, Commitment, CommitmentSignature)'. We need a 'ProtocolMagic'
+-- to make the 'CommitmentSignature ~ Signature (EpochIndex, Commitment)`
+genCommitmentsMap :: ProtocolMagic -> Gen CommitmentsMap
+genCommitmentsMap pm = mkCommitmentsMap <$> listOf (genSignedCommitment pm)
```

Comment:
This is duplicated in 3 places! Here, `core/test/Test/Pos/Core/Gen.hs`, and `ssc/test/Test/Pos/Ssc/Arbitrary.hs`.

I'd like to understand why, and if it's just a temporary stopgap. Couldn't `core-test`'s functions/instances just be imported by the packages "above" it?

Reply:
Now its only two versions. The ones in `core/test/Test/Pos/Core/Gen.hs` use Hedgehog's `Gen` type where as the `Arbitrary` ones use `QuickCheck`'s `Gen` which are actually different types. So we actually need both for now.

Reply:
Ah gotcha, didn't realize they were different types. But couldn't the `infra` occurrence just import the `core` one?